xline/src/server/xline_server.rs

```diff
@@ -285,124 +238,133 @@ where
+            AuthServer::new(
+                auth_storage,
+                client,
+                self.cluster_info.get_self_id().clone(),
+            ),
+            WatchServer::new(watcher, Arc::clone(&header_gen)),
+            MaintenanceServer::new(persistent, header_gen),
+            Self::construct_curp_server(
+                Arc::clone(&self.cluster_info),
+                self.is_leader,
+                ce,
+                state,
+                &self.storage_cfg,
+                Arc::clone(&self.curp_cfg),
+            )
+            .await,
+        ))
+    }
+
+    /// Construct a new `CurpServer`
+    async fn construct_curp_server<S: StorageApi, CE: CurpCommandExecutor<Command> + 'static>(
+        cluster_info: Arc<ClusterMember>,
+        is_leader: bool,
+        ce: CE,
+        state: State<S>,
+        storage_cfg: &StorageConfig,
+        curp_cfg: Arc<CurpConfig>,
+    ) -> CurpServer<S> {
+        match *storage_cfg {
             StorageConfig::Memory => {
                 CurpServer::new(
-                    self.id(),
-                    self.is_leader(),
-                    self.state.others(),
-                    CommandExecutor::new(
-                        Arc::clone(&self.kv_storage),
-                        Arc::clone(&self.auth_storage),
-                        Arc::clone(&self.lease_storage),
-                        Arc::clone(&self.persistent),
-                        Arc::clone(&self.index_barrier),
-                        Arc::clone(&self.id_barrier),
-                        self.header_gen.revision_arc(),
-                        Arc::clone(&self.auth_revision),
-                    ),
+                    cluster_info,
+                    is_leader,
+                    ce,
                     MemorySnapshotAllocator,
-                    Arc::clone(&self.curp_cfg),
+                    state,
+                    Arc::clone(&curp_cfg),
                     None,
                 )
                 .await
             }
             StorageConfig::RocksDB(_) => {
                 CurpServer::new(
-                    self.id(),
-                    self.is_leader(),
-                    self.state.others(),
-                    CommandExecutor::new(
-                        Arc::clone(&self.kv_storage),
-                        Arc::clone(&self.auth_storage),
-                        Arc::clone(&self.lease_storage),
-                        Arc::clone(&self.persistent),
-                        Arc::clone(&self.index_barrier),
-                        Arc::clone(&self.id_barrier),
-                        self.header_gen.revision_arc(),
-                        Arc::clone(&self.auth_revision),
-                    ),
+                    cluster_info,
+                    is_leader,
+                    ce,
                     RocksSnapshotAllocator,
-                    Arc::clone(&self.curp_cfg),
+                    state,
+                    Arc::clone(&curp_cfg),
```

Comment:
I suggesst change the snapshot_alloctor to a dynamic type like `Box<dyn xxx>`, which can further reduce duplication here

Reply:
IMO, we can maintain the status quo in that using dynamic type would have some performance cost.  And there is no point to keep the `construct_curp_server` function here if we use a dynamic type.

Reply:
we already use it as a dynamic type now, it will be boxed in `CurpNode`, my suggestion is just box it early, will not bring performance cost

Reply:
Done